File: Data Structures & Algorithms/longest-repeating-substring-with-replacement/submission-13.cpp
class Solution {
public:
    int characterReplacement(string str, int k) {

        int longest_substr = 0;
        int left = 0;
        int most_freq_occur_char = 0;
        std::unordered_map<char, int> freq;

        for (int right = 0; right < str.size(); right++){
            // add to the map
            freq[str[right]] += 1; 
            
            // check if the thing we added is now the most freq occuring char
            // leverage the fact that the most freq occuring char only changes upon adding
            most_freq_occur_char = std::max(most_freq_occur_char, freq[str[right]]); 

            while ((right - left + 1) - most_freq_occur_char > k){
                freq[str[left++]] -= 1;
            }

            longest_substr = std::max(longest_substr, right - left + 1);
        }


        return longest_substr;
    }
};

File: Data Structures & Algorithms/longest-substring-without-duplicates/submission-13.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:

        left, right, max_substring = 0, 0, 0
        substring = set()
        
        while right < len(s):
            max_substring = max(max_substring, len(substring))
            while s[right] in substring:
                substring.remove(s[left])
                left += 1
            substring.add(s[right])
            right += 1
        return max(max_substring, len(substring))